BB2-342 Improve token and request_logging

dtisza1 opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BB2-342](https://jira.cms.gov/browse/BB2-342)


**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
As a BB Engineer, I need improved BB20 application logging so that I can more easily audit the system and peripheral systems activity.

Summary:

This improves logging items related to token endpoint activity. This is done mainly via the `hhs_oauth_server/request_logging.py` custom logging Django middleware.

In addition, various other useful items were added to the logging. See the key legend comments in `hhs_oauth_server/request_logging.py` for a full listing.

An enviornment (ENV) type variable `DJANGO_LOG_JSON_FORMAT_PRETTY` was added to enable "pretty printed" JSON logging when `True`.  This helps to make the logging more readable for local development.

The logging unit tests have been refactored in `apps/logging/tests/test_audit_loggers.py` to utilize [JSON Schema](https://json-schema.org/) for validation. This is much easier to read and more versatile!

Schema definitions for the various types of logging events are in `apps/logging/tests/audit_logger_schemas.py`.

There is a bug fix included related to the integration with SLSx PR https://github.com/CMSgov/bluebutton-web-server/pull/866.  The "state" parameter was being used in the callback to pick back up the auth flow UUID info from the previous session. For the SLSx flow, this parameter was changed to "relay".  So this PR should be deployed before we integrate with SLSx.

### What Does This PR Do?

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->
**The following is a list of changes from the commits (including additional comments):**

* Add ENV option to pretty print json for local dev 
  - Add setting `DJANGO_LOG_JSON_FORMAT_PRETTY=True` to docker-compose.yml
  - Add setting of LOG_JSON_FORMAT_PRETTY based on ENV var DJANGO_LOG_JSON_FORMAT_PRETTY in hhs_oauth_server/settings/base.py. This is to be used for pretty printing JSON logging for easier reading in local development.

* Add JSON logging pretty print to loggers
  - Add pretty print to logging in apps/fhir/server/loggers.py, apps/logging/signals.py, apps/logging/serializers.py, apps/mymedicare_cb/loggers.py
  - Add crosswalk to Token serializer in apps/logging/serializers.py
  - Sort "type" field to top of logging items for serialized objects in apps/logging/serializers.py.

* Refactor hhs_oauth_server/request_logging.py
  - Add "type" label field as "request_response_middleware"
  - Add private methods for updating the self.log_msg[] dictionary: _log_msg_update_from_dict(), _log_msg_update_from_object(), _log_msg_update_from_querydict()
  - Init some empty fields for dashboards compatibility in RequestResponseLog.__init__()
  - Add logging items from request.POST
  - Add logging items from request.headers
  - Add logging items from request.body.  Including getting application info from client_id, and info related to token_refresh.
  - Add logging items from request.GET for query params
  - Add logging items from request access token
  - Add logging items from response content (refresh_token)
  - Get access token hash info prior to refresh in RequestTimeLoggingMiddleware.process_request()
  - Update listing of logging fields in key/legend comments.

* Refactor logging tests to use JSON Schema validation
  - Create schemas for validating logging entries in apps/logging/tests/audit_logger_schemas.py
  - Create _validateJsonSchema() method for use in tests. This utilizes jsonschema.validate() to check JSON content against a schema.  in apps/logging/tests/test_audit_loggers.py.
  - Update tests to use JSON Schema validation in apps/logging/tests/test_audit_loggers.py.  NOTE: Did not update tests in test_callback_url_success_sls_logger() since these will be removed after the SLSx migration.


### What Should Reviewers Watch For?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->
If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->

NOTE: This PR includes a bug fix for the auth flow log tracing when integrating with the new SLSx. This PR should be deployed before we perform the cut-over to SLSx!!!

### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [ X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X ] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X ] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

njdister commented on hhs_oauth_server/request_logging.py:
```diff
@@ -20,96 +23,336 @@
+        - req_qparam_count = Info from HTTP Query Params.
+        - req_qparam_format = Info from HTTP Query Params.
+        - req_qparam__id = Info from HTTP Query Params.
+        - req_qparam_lastupdated = Info from HTTP Query Params.
+        - req_qparam_patient = Info from HTTP Query Params.
+        - req_qparam_response_type = Info from HTTP Query Params.
+        - req_qparam_startindex = Info from HTTP Query Params.
+        - req_qparam_type = Info from HTTP Query Params.
+        - req_redirect_uri = Request redirect URI.
+        - req_refresh_token_hash = Request refresh token hash.
+        - req_response_type = Response type.
+        - req_scope = Request scope
+        - req_share_demographic_scopes = Request bene share demo scopes.
+        - request_method = http request method type (GET/POST/etc)
+        - request_post_grant_type = Post grant type.
+        - request_scheme = The scheme of the request (http/https/etc)
+        - request_uuid = The UUID identifying the request.
+        - req_user_id = Request user id.
+        - req_user_username = Request username.
+        - resp_access_token_hash = Response AC info.
+        - resp_access_token_scopes = Response AC info.
+        - resp_app_id = Response AC info.
+        - resp_app_name = Response AC info.
+        - resp_app_require_demographic_scopes = Response AC info.
+        - resp_dev_id = Response AC info.
+        - resp_dev_name = Response AC info.
+        - resp_expires_in = Response AC info.
+        - resp_fhir_id = Response AC info.
+        - response_code = The response status code.
+        - resp_refresh_token_hash = Response refresh token info.
+        - resp_scope = Response refresh token info.
+        - resp_token_type = Response refresh token info.
+        - resp_user_id = Response refresh token info.
+        - resp_user_username = Response refresh token info.
+        - size = Size in bytes of the response.content
+        - start_time = Unix Epoch format time of the request processed.
+        - type = Label for type of log. Set to "request_response_middleware".
+        - user_id = Login user (or None) or OAuth2 API id.
+        - user = Login user (or None) or OAuth2 API username.
+        - user_username = Login user (or None) or OAuth2 API username.
```
It would be good to document which fields are legacy fields that retained their name for compatibility, to delineate from `request_` vs `req_`, etc.

dtisza1 replied:
@njdister I updated the comments with a "BB2-342" tag next to the newly added field names.
Also fixed a bug I introduced while renaming "request_post_grant_type" --> "req_post_grant_type".

njdister reviewed (approved):
Excellent work with this!  The pretty printing is very nice for local development, jsonschema is a much easier way of unit testing our logs, and the extra fields will be very useful for future insights/dashboards.

As discussed we should deploy this branch to a sandbox canary attached to a real load balancer for testing, just to be safe.

dtisza1 commented:
> Excellent work with this! The pretty printing is very nice for local development, jsonschema is a much easier way of unit testing our logs, and the extra fields will be very useful for future insights/dashboards.
> 
> As discussed we should deploy this branch to a sandbox canary attached to a real load balancer for testing, just to be safe.
@njdister I like this idea for testing!

JFU-GIT reviewed (approved):
Tested locally, pretty print option makes it much easier for human readers, works as expected.

dtisza1 commented:
Reviewed the logging in Splunk after Nathan setup an instance in our sandbox.  The logging is looking good and not producing any errors.  I also found some additional items to log and added to this PR.  Also fixed a bug with auth flow trace logging and SLSx.  We will want to deploy this before migrating to SLSx/Medicare.gov.

I believe this is ready and working as expected!

dtisza1 commented:
I reviewed the logging from the one instance in Sandbox running this PR's branch. Everything is still looking good. New items are showing in Splunk. In New Relic, I'm not seeing any new types of errors or errors related to the new logging additions.

This looks good, so merging this to master next.
